### PR TITLE
test(e2e/twap): twap query before/after swap and after pruning

### PR DESCRIPTION
**Pull request by p0mvn**
Closes: #XXX

## What is the purpose of the change

Basic e2e testing of TWAP query before/after swap and after pruning

## Testing and Verifying

This change is a trivial rework / code cleanup without any test coverage.

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? no
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? no
  - How is the feature or change documented? not applicable

### CODE REVIEW

File: tests/e2e/e2e_test.go
Comment (p0mvn) [inline tests/e2e/e2e_test.go:220]: I'm currently trying to figure out why these would not be equal.

My thinking is that they should still be since they are supposed to utilize the same start record

Comment (p0mvn) [reply tests/e2e/e2e_test.go:220]: Ah right this should be addressed in https://github.com/osmosis-labs/osmosis/pull/2635

Comment (ValarDragon) [reply tests/e2e/e2e_test.go:220]: Yeah, we have a bug if they're not equal. (Though this line uses time.Now() right above)

Comment (p0mvn) [reply tests/e2e/e2e_test.go:220]: You are right, `time.Now()` was the problem, thanks. The new check you added exposed the invalid setup. Then, making time be at fixed intervals from queried block times helped fix this.

All tests are passing right now. I'm going to make a 3-asset pool next to try and repro the bug found in the review

File: tests/e2e/configurer/chain/chain.go
Comment (p0mvn) [inline tests/e2e/configurer/chain/chain.go:95]:
```suggestion
// WaitForEpochs waits for the chain to go through a given number of epochs.
```

File: tests/e2e/e2e_test.go
Comment (p0mvn) [inline tests/e2e/e2e_test.go:261]: This test sometimes fails and sometimes passes. However, I've verified through logs that it's grabbing the records from the desired heights that have correct spot prices.

The problem is likely coming from either time deltas or accumulator calculations when interpolating records.

@ValarDragon how do you think I should go about this? Is it worth investigating further or can we attribute this to setup/rounding errors? 

I might also be doing something wrong. Spent hours going in circles trying to root cause what exactly

File: tests/e2e/e2e_test.go
Comment (p0mvn) [inline tests/e2e/e2e_test.go:224]: This check is the same. Fails sometimes.

The records are as expected, only accumulators and end times are different

File: tests/e2e/e2e_test.go
Comment (p0mvn) [inline tests/e2e/e2e_test.go:230]: Note to reviewers:

I've been debugging this with extra logs added.

I observed that records are picked up from the correct heights and with the correct spot prices.

I also logged time deltas and accumulator differences from:
https://github.com/osmosis-labs/osmosis/blob/7a2db79f406adfd3a5099e4acb1b1936b6e25c06/x/twap/types/utils.go#L55-L56

It seems that these are not exactly equal in cases when the time deltas are different. So I suspect that these are rounding issues caused by truncations.

Another potential source is sig fig round for spot prices. I attempted to increase precision by 4 sig figs for testing purposes. Although it reduced the delta, it did not completely eliminate it.

File: tests/e2e/configurer/chain/node.go
Comment (ValarDragon) [inline tests/e2e/configurer/chain/node.go:108]: Not blocking this PR, but generally confused about direction to not return an error here to flag in calling test, vs logging an error.

File: tests/e2e/e2e_test.go
Comment (czarcas7ic) [inline tests/e2e/e2e_test.go:226]: I might be missing something but this check seems redundant, i.e. if the previous check passes there really should be no logical way this would fail. What do you think?

File: tests/e2e/e2e_test.go
Comment (czarcas7ic) [inline tests/e2e/e2e_test.go:214]: While I don't think we should be checking for exact values, I do think NotEqual might be slightly too broad here (for example, a nil record is not equal but probably not what we want to see). I think a safe middle ground would just be doing some napkin math and at least asserting one should be GT or LT the other. Thoughts?

File: tests/e2e/e2e_test.go
Comment (czarcas7ic) [inline tests/e2e/e2e_test.go:254]:
```suggestion
	s.Require().ErrorContains(err, "too old")
```
nit: might be redundant

File: tests/e2e/e2e_test.go
Comment (czarcas7ic) [inline tests/e2e/e2e_test.go:260]: I think both of these records are before pruning right? The comment and var name implies you are trying to test one before and one after pruning but I feel like they are both pre pruning.

**Review by czarcas7ic — APPROVED** [review]
Well done on this! I had to reread some of these quite a few times to ensure I was tracking, but I think the way you named them and the comments you added made it as clear as possible. 

After these comments are addressed I think its good to merge 👍

File: tests/e2e/configurer/chain/node.go
Comment (p0mvn) [reply tests/e2e/configurer/chain/node.go:108]: `Errorf` does not only log but also fails immediately.

The reason for changing this is to reduce verbosity and simplify tests.

For the majority of these methods, we fail if the error is non-nil anyways. However, constantly doing the error checks reduces the readability of the tests.

Failing immediately via this `testing` package method reduces verbosity and makes it easier to write tests

File: tests/e2e/e2e_test.go
Comment (p0mvn) [reply tests/e2e/e2e_test.go:214]: In my opinion, the exact math should be tested at a lower level of abstraction. In e2e, we are just checking that system components work well together without diving deep into the calculation details.

Also, this is not 100% deterministic since this is twap to now where blocks and their respective time might slightly vary between runs

File: tests/e2e/e2e_test.go
Comment (p0mvn) [reply tests/e2e/e2e_test.go:226]: This is `QueryGetArithmeticTwap` while previous one is `QueryGetArithmeticTwapToNow`. The goal here was to test that both are functioning as expected

File: tests/e2e/e2e_test.go
Comment (p0mvn) [reply tests/e2e/e2e_test.go:259]: - `twapFromAfterToAfterSwapAndAfterPruning` is after pruning
   * renamed to `twapAfterPruning10ms`
- `twapFromAfterToAfterSwapAndBeforePruning` is before pruning
   * renamed to `twapAfterSwapBeforePruning10Ms`

renamed both for clarity. Let me know what you think

File: tests/e2e/e2e_test.go
Comment (czarcas7ic) [reply tests/e2e/e2e_test.go]: Ah I didn't notice that, this makes sense then

File: tests/e2e/e2e_test.go
Comment (czarcas7ic) [reply tests/e2e/e2e_test.go]: This is clearer to me, thanks!

File: tests/e2e/e2e_test.go
Comment (czarcas7ic) [reply tests/e2e/e2e_test.go:214]: I agree that the exact math should not be tested here. I guess what I am saying is `NotEqual` will return True if the twap record we retrieve is nil. We should for sure know which twap (before or after swap) is greater than the other, so a more specific `s.Require().True(twapFromBeforeSwapToBeforeSwapOne.GT(twapFromBeforeSwapToAfterSwap))` (or whatever the properly verbiage is) might be more desirable. 

This is not PR blocking at all, just a suggestion that may save us from missing issues in the future.

File: tests/e2e/e2e_test.go
Comment (ValarDragon) [inline tests/e2e/e2e_test.go:192]: This may not be reliable in edge cases due to rounding concerns

**Review by ValarDragon — APPROVED** [review]

File: tests/e2e/e2e_test.go
Comment (p0mvn) [reply tests/e2e/e2e_test.go:192]: I haven't seen this one failing but I will add the approximate equality check just in case

File: tests/e2e/e2e_test.go
Comment (p0mvn) [reply tests/e2e/e2e_test.go]: Will try to address this before merging. Marking this as draft for now

**Pull request merged**
